src/test/java/org/jenkinsci/plugins/workflow/graphanalysis/TestFlowScanner.java

```diff
@@ -0,0 +1,428 @@
+            System.out.println("Testing class: "+sa.getClass());
+            FlowNode node = sa.findFirstMatch(heads, null, (Predicate)Predicates.alwaysFalse());
+            Assert.assertNull(node);
+
+            Collection<FlowNode> nodeList = sa.filteredNodes(heads, null, (Predicate) Predicates.alwaysFalse());
+            Assert.assertNotNull(nodeList);
+            Assert.assertEquals(0, nodeList.size());
+        }
+
+
+        CollectingVisitor vis = new CollectingVisitor();
+        // Verify we touch head and foot nodes too
+        for (AbstractFlowScanner sa : scans) {
+            System.out.println("Testing class: " + sa.getClass());
+            Collection<FlowNode> nodeList = sa.filteredNodes(heads, null, (Predicate) Predicates.alwaysTrue());
+            vis.reset();
+            sa.visitAll(heads, vis);
+            Assert.assertEquals(5, nodeList.size());
+            Assert.assertEquals(5, vis.getVisited().size());
+        }
+
+        // Test with a stop node given, sometimes no matches
+        Collection<FlowNode> noMatchEndNode = Collections.singleton(exec.getNode("5"));
+        Collection<FlowNode> singleMatchEndNode = Collections.singleton(exec.getNode("4"));
+        for (AbstractFlowScanner sa : scans) {
+            FlowNode node = sa.findFirstMatch(heads, noMatchEndNode, MATCH_ECHO_STEP);
+            Assert.assertNull(node);
+
+            Collection<FlowNode> nodeList = sa.filteredNodes(heads, noMatchEndNode, MATCH_ECHO_STEP);
+            Assert.assertNotNull(nodeList);
+            Assert.assertEquals(0, nodeList.size());
+
+            // Now we try with a stop list the reduces node set for multiple matches
+            node = sa.findFirstMatch(heads, singleMatchEndNode, MATCH_ECHO_STEP);
+            Assert.assertEquals(exec.getNode("5"), node);
+            nodeList = sa.filteredNodes(heads, singleMatchEndNode, MATCH_ECHO_STEP);
+            Assert.assertNotNull(nodeList);
+            Assert.assertEquals(1, nodeList.size());
+            Assert.assertEquals(exec.getNode("5"), nodeList.iterator().next());
+        }
+    }
+
+    /** Tests the basic scan algorithm where blocks are involved */
+    @Test
+    public void testBasicScanWithBlock() throws Exception {
+        WorkflowJob job = r.jenkins.createProject(WorkflowJob.class, "Convoluted");
+        job.setDefinition(new CpsFlowDefinition(
+            "echo 'first'\n" +
+            "timeout(time: 10, unit: 'SECONDS') {\n" +
+            "    echo 'second'\n" +
+            "    echo 'third'\n" +
+            "}\n" +
+            "sleep 1"
+        ));
+        /** Flow structure (ID - type)
+         2 - FlowStartNode
+         3 - EchoStep
+         4 - TimeoutStep
+         5 - TimeoutStep with BodyInvocationAction
+         6 - EchoStep
+         7 - EchoStep
+         8 - StepEndNode (BlockEndNode), startId=5
+         9 - StepEndNode (BlockEndNode), startId = 4
+         10 - SleepStep
+         11 - FlowEndNode
```

Comment:
This is a really good notion; I need to mull a bit on testing because we need to ensure we have robust coverage on the many flow graph structures.

To be honest, we really need a robust mocking framework for flow graphs -- the one used in stage view is handy, but very limited to the stage-specific use case.